File: src/gps_node/src/gps_node.cpp
#include <iostream>
#include <string>
#include <sstream>
#include <vector>
#include <iomanip>
#include <cstdlib>
#include <cstring>
#include <fcntl.h>
#include <unistd.h>
#include <termios.h>
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/nav_sat_fix.hpp>
#include <sensor_msgs/msg/nav_sat_status.hpp>


// RMC → position, speed, heading
// GGA → position, altitude, satellite count  
// GSA → fix quality, HDOP, PDOP (accuracy metrics)

struct GGA {
    std::string time;
    double lat, lon;
    std::string fix, sats, alt, alt_unit;
};

struct RMC {
    std::string time, status;
    double lat, lon;
    std::string speed, course, date;
};

struct GSA {
    std::string fix_mode, pdop, hdop, vdop;
};

enum class SentenceType { GGA, RMC, GSA, UNKNOWN };

double nmea_to_decimal(const std::string& coord, const std::string& dir) {
    if (coord.empty()) return 0.0;
    auto dot = coord.find('.');
    if (dot == std::string::npos || dot < 2) return 0.0;
    double degrees = std::stod(coord.substr(0, dot - 2));
    double minutes = std::stod(coord.substr(dot - 2));
    double decimal = degrees + minutes / 60.0;
    if (dir == "S" || dir == "W") decimal = -decimal;
    return decimal;
}

std::string strip_checksum(const std::string& s) {
    auto star = s.find('*');
    return (star != std::string::npos) ? s.substr(0, star) : s;
}

const std::string& field(const std::vector<std::string>& f, size_t i) {
    static const std::string empty;
    return i < f.size() ? f[i] : empty;
}

SentenceType get_sentence_type(const std::string& id) {
    std::string s = strip_checksum(id);
    if (!s.empty() && s[0] == '$') s = s.substr(1);
    if (s.size() < 3) return SentenceType::UNKNOWN;
    std::string tag = s.substr(s.size() - 3);

    if (tag == "GGA") return SentenceType::GGA;
    if (tag == "RMC") return SentenceType::RMC;
    if (tag == "GSA") return SentenceType::GSA;
    return SentenceType::UNKNOWN;
}

std::vector<std::string> split(const std::string& line, char delim) {
    std::vector<std::string> fields;
    std::stringstream ss(line);
    std::string f;
    while (std::getline(ss, f, delim))
        fields.push_back(f);
    return fields;
}

GGA parse_gga(const std::vector<std::string>& f) {
    return {
        field(f, 1),
        nmea_to_decimal(field(f, 2), field(f, 3)),
        nmea_to_decimal(field(f, 4), field(f, 5)),
        field(f, 6), field(f, 7), field(f, 9), field(f, 10)
    };
}

RMC parse_rmc(const std::vector<std::string>& f) {
    return {
        field(f, 1), field(f, 2),
        nmea_to_decimal(field(f, 3), field(f, 4)),
        nmea_to_decimal(field(f, 5), field(f, 6)),
        field(f, 7), field(f, 8), field(f, 9)
    };
}

GSA parse_gsa(const std::vector<std::string>& f) {
    return {
        field(f, 2),
        field(f, 15), field(f, 16),
        strip_checksum(field(f, 17))
    };
}


int open_serial(const std::string& port) {
    int fd = open(port.c_str(), O_RDONLY | O_NOCTTY);
    if (fd < 0) {
        std::cerr << "Failed to open " << port << ": " << strerror(errno) << "\n";
        return -1;
    }

    termios tty{};
    if (tcgetattr(fd, &tty) != 0) {
        std::cerr << "tcgetattr failed: " << strerror(errno) << "\n";
        close(fd);
        return -1;
    }

    cfsetispeed(&tty, B9600);
    cfsetospeed(&tty, B9600);

    tty.c_cflag = (tty.c_cflag & ~CSIZE) | CS8;
    tty.c_cflag &= ~(PARENB | CSTOPB | CRTSCTS);
    tty.c_cflag |= CLOCAL | CREAD;

    tty.c_iflag &= ~(IXON | IXOFF | IXANY | IGNBRK | BRKINT | PARMRK | ISTRIP | INLCR | IGNCR | ICRNL);
    tty.c_lflag &= ~(ECHO | ECHONL | ICANON | ISIG | IEXTEN);
    tty.c_oflag &= ~OPOST;

    tty.c_cc[VMIN] = 1;
    tty.c_cc[VTIME] = 0;

    if (tcsetattr(fd, TCSANOW, &tty) != 0) {
        std::cerr << "tcsetattr failed: " << strerror(errno) << "\n";
        close(fd);
        return -1;
    }

    return fd;
}

std::string read_line(int fd) {
    std::string line;
    char c;
    while (true) {
        ssize_t n = read(fd, &c, 1);
        if (n < 0) {
            std::cerr << "Read error: " << strerror(errno) << "\n";
            return "";
        }
        if (n == 0) continue;
        if (c == '\n') return line;
        if (c != '\r') line += c;
    }
}

class GpsNode : public rclcpp::Node {
    public:
        GpsNode() : Node("gps_node") {
            fix_publisher_ = this->create_publisher<sensor_msgs::msg::NavSatFix>("/gps/fix", 10);
            status_publisher_ = this->create_publisher<sensor_msgs::msg::NavSatStatus>("/gps/status", 10);


            const char* env_port = std::getenv("GPS_PORT");
            std::string port = env_port ? env_port : "/dev/ttyUSB0";

            fd_ = open_serial(port);

            timer_ = this->create_wall_timer(
                std::chrono::milliseconds(100),
                std::bind(&GpsNode::loop, this)
            );

            RCLCPP_INFO(this->get_logger(), "GPS node started on %s", port.c_str());
        }

    private:
        void loop() {
            if (fd_ < 0) return;
            std::string line = read_line(fd_);
            if (line.empty() || line[0] != '$') return;

            auto fields = split(line, ',');
            if (fields.empty()) return;

            switch (get_sentence_type(fields[0])) {
                case SentenceType::GGA: publishGGA(parse_gga(fields)); break;
                case SentenceType::RMC: publishRMC(parse_rmc(fields)); break;
                case SentenceType::GSA: publishGSA(parse_gsa(fields)); break;
                default: break;
            }
        }

        void publishRMC(const RMC& r) {
            if (r.status != "A") return;
            auto msg = sensor_msgs::msg::NavSatFix();
            msg.latitude = r.lat;
            msg.longitude = r.lon;
            msg.status.status = sensor_msgs::msg::NavSatStatus::STATUS_FIX;
            fix_publisher_->publish(msg);
            RCLCPP_INFO(this->get_logger(), "[RMC] Lat: %.6f  Lon: %.6f  Speed: %s kn", r.lat, r.lon, r.speed.c_str());
        }

        void publishGGA(const GGA& g) {
            if (g.fix == "0") return;
            auto msg = sensor_msgs::msg::NavSatFix();

            msg.latitude = g.lat;
            msg.longitude = g.lon;
            msg.altitude = g.alt.empty() ? 0.0 : std::stod(g.alt);
            msg.status.status = sensor_msgs::msg::NavSatStatus::STATUS_FIX;
            fix_publisher_->publish(msg);
            RCLCPP_INFO(this->get_logger(), "[GGA] Lat: %.6f  Lon: %.6f  Alt: %s m  Sats: %s", g.lat, g.lon, g.alt.c_str(), g.sats.c_str());
        }

        void publishGSA(const GSA& g) {
            auto msg = sensor_msgs::msg::NavSatStatus();
            msg.status = (g.fix_mode == "1") ? 
                sensor_msgs::msg::NavSatStatus::STATUS_NO_FIX :
                sensor_msgs::msg::NavSatStatus::STATUS_FIX;
            status_publisher_->publish(msg);
            RCLCPP_INFO(this->get_logger(), "[GSA] Fix: %s  HDOP: %s  PDOP: %s", g.fix_mode.c_str(), g.hdop.c_str(), g.pdop.c_str());
        }

        rclcpp::Publisher<sensor_msgs::msg::NavSatFix>::SharedPtr fix_publisher_;
        rclcpp::Publisher<sensor_msgs::msg::NavSatStatus>::SharedPtr status_publisher_;
        rclcpp::TimerBase::SharedPtr timer_;
        int fd_ = -1;
};


int main(int argc, char* argv[]) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<GpsNode>());
    rclcpp::shutdown();
    return 0;
};

File: src/gps_node/gps_node/gps_node.py
import os
import termios

import rclpy
from rclpy.node import Node
from sensor_msgs.msg import NavSatFix, NavSatStatus


def nmea_to_decimal(coord: str, direction: str) -> float:
    if not coord:
        return 0.0

    dot = coord.find('.')
    if dot == -1 or dot < 2:
        return 0.0

    degrees = float(coord[: dot - 2])
    minutes = float(coord[dot - 2 :])
    decimal = degrees + minutes / 60.0
    if direction in {'S', 'W'}:
        decimal = -decimal
    return decimal


def strip_checksum(value: str) -> str:
    star = value.find('*')
    return value[:star] if star != -1 else value


def sentence_type(identifier: str) -> str:
    stripped = strip_checksum(identifier)
    if stripped.startswith('$'):
        stripped = stripped[1:]
    if len(stripped) < 3:
        return 'UNKNOWN'
    return stripped[-3:]


class GpsNode(Node):
    def __init__(self):
        super().__init__('gps_node')

        self.drone_id = self.declare_parameter(
            'drone_id',
            os.getenv('AETHER_DRONE_ID', 'AE-01'),
        ).value
        self.port = self.declare_parameter(
            'gps_port',
            os.getenv('GPS_PORT', '/dev/ttyUSB0'),
        ).value
        self.baud_rate = int(
            self.declare_parameter(
                'baud_rate',
                int(os.getenv('AETHER_GPS_BAUD_RATE', '9600')),
            ).value
        )
        self.fix_topic = self.declare_parameter(
            'fix_topic',
            f'/{self.drone_id}/gps/fix',
        ).value
        self.status_topic = self.declare_parameter(
            'status_topic',
            f'/{self.drone_id}/gps/status',
        ).value

        self.fix_publisher = self.create_publisher(NavSatFix, self.fix_topic, 10)
        self.status_publisher = self.create_publisher(
            NavSatStatus,
            self.status_topic,
            10,
        )

        self.serial_file = None
        self.connect_serial()
        self.timer = self.create_timer(0.1, self.loop)
        self.get_logger().info(
            f'GPS node started for {self.drone_id} on {self.port}'
        )

    def connect_serial(self):
        try:
            raw = open(self.port, 'rb', buffering=0)
            attrs = termios.tcgetattr(raw.fileno())
            attrs[0] &= ~(
                termios.IXON
                | termios.IXOFF
                | termios.IXANY
                | termios.IGNBRK
                | termios.BRKINT
                | termios.PARMRK
                | termios.ISTRIP
                | termios.INLCR
                | termios.IGNCR
                | termios.ICRNL
            )
            attrs[1] &= ~termios.OPOST
            attrs[2] &= ~(termios.PARENB | termios.CSTOPB | termios.CSIZE | termios.CRTSCTS)
            attrs[2] |= termios.CLOCAL | termios.CREAD | termios.CS8
            attrs[3] &= ~(termios.ECHO | termios.ECHONL | termios.ICANON | termios.ISIG | termios.IEXTEN)
            attrs[6][termios.VMIN] = 1
            attrs[6][termios.VTIME] = 0

            speed_attr = self.baud_constant(self.baud_rate)
            termios.cfsetispeed(attrs, speed_attr)
            termios.cfsetospeed(attrs, speed_attr)
            termios.tcsetattr(raw.fileno(), termios.TCSANOW, attrs)
            self.serial_file = raw
        except Exception as exc:
            self.serial_file = None
            self.get_logger().error(f'Failed to open GPS serial port {self.port}: {exc}')

    def baud_constant(self, baud_rate: int):
        mapping = {
            4800: termios.B4800,
            9600: termios.B9600,
            19200: termios.B19200,
            38400: termios.B38400,
            57600: termios.B57600,
            115200: termios.B115200,
        }
        return mapping.get(baud_rate, termios.B9600)

    def loop(self):
        if self.serial_file is None:
            return

        try:
            line = self.serial_file.readline().decode('ascii', errors='ignore').strip()
        except Exception as exc:
            self.get_logger().error(f'GPS read error: {exc}')
            return

        if not line or not line.startswith('$'):
            return

        fields = line.split(',')
        if not fields:
            return

        message_type = sentence_type(fields[0])
        if message_type == 'GGA':
            self.publish_gga(fields)
        elif message_type == 'RMC':
            self.publish_rmc(fields)
        elif message_type == 'GSA':
            self.publish_gsa(fields)

    def publish_rmc(self, fields):
        status = fields[2] if len(fields) > 2 else ''
        if status != 'A':
            return

        msg = NavSatFix()
        msg.latitude = nmea_to_decimal(fields[3] if len(fields) > 3 else '', fields[4] if len(fields) > 4 else '')
        msg.longitude = nmea_to_decimal(fields[5] if len(fields) > 5 else '', fields[6] if len(fields) > 6 else '')
        msg.status.status = NavSatStatus.STATUS_FIX
        self.fix_publisher.publish(msg)

    def publish_gga(self, fields):
        fix = fields[6] if len(fields) > 6 else '0'
        if fix == '0':
            return

        msg = NavSatFix()
        msg.latitude = nmea_to_decimal(fields[2] if len(fields) > 2 else '', fields[3] if len(fields) > 3 else '')
        msg.longitude = nmea_to_decimal(fields[4] if len(fields) > 4 else '', fields[5] if len(fields) > 5 else '')
        msg.altitude = float(fields[9]) if len(fields) > 9 and fields[9] else 0.0
        msg.status.status = NavSatStatus.STATUS_FIX
        self.fix_publisher.publish(msg)

    def publish_gsa(self, fields):
        fix_mode = fields[2] if len(fields) > 2 else '1'
        msg = NavSatStatus()
        msg.status = (
            NavSatStatus.STATUS_NO_FIX
            if fix_mode == '1'
            else NavSatStatus.STATUS_FIX
        )
        self.status_publisher.publish(msg)


def main(args=None):
    rclpy.init(args=args)
    node = GpsNode()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == '__main__':
    main()
